crates/pallet-domains/src/lib.rs

```diff
@@ -810,14 +809,18 @@ mod pallet {
             origin: OriginFor<T>,
             runtime_name: Vec<u8>,
             runtime_type: RuntimeType,
-            code: Vec<u8>,
+            raw_genesis_storage: Vec<u8>,
```

Comment:
From what I recall raw genesis storage was a JSON string, why is it a vector of bytes here (and in other places too) and not a string? JSON is not a binary format. I think I already commented about this in one of the other PRs.

Reply:
I think we should change it to string everywhere we have it then because it is a string.

Reply:
I see it is decoded from raw bytes later, can't we take the decoded instance here as an argumnet then? Otherwise we're encoding/decoding it back and forth unnecessarily. Also I just noticed `runtime_name` is a vector of bytes for some reason rather than a string too :confused:

Reply:
Do you mean `raw_genesis_storage: RawGenesis`?  that may not be friendly for those who want to initiate this call through `polkadot.js` or something similar, with `Vec<u8>` the user can upload a file directly.

Will change `runtime_name` and `domain_name` to string in later commit.

Reply:
Are you sure they can't upload the file directly with other data types? I think at least when implementing some trait on it it should be possible to do so.

Reply:
I think it will treat `RawGenesis` as `struct` data type and request to provide each of its fields, and I just verified:

with `raw_genesis_storage: Vec<u8>`:
<img width="825" alt="截屏2023-09-15 06 30 39" src="https://github.com/subspace/subspace/assets/28501710/af975ab1-23ee-41ee-989c-44798a94ca0a">

with `raw_genesis_storage: RawGenesis`:
<img width="822" alt="截屏2023-09-15 06 31 06" src="https://github.com/subspace/subspace/assets/28501710/71395d54-9f57-4faf-b846-8ee3cdb07f2c">

Reply:
Hm, I see, but I'm still wondering if we can provide some kind of metadata to make upload work there like you expect it to

Reply:
> if we can provide some kind of metadata to make upload work there like you expect it to

I didn't find this after a few searches and gave up but a TODO was added

>  Also I just noticed runtime_name is a vector of bytes for some reason rather than a string too 😕

I tried to use `String` in runtime but `sp_std::alloc::String` is not supported, I did try `scale_info::prelude::string::String` and got it complied but it does not seem legitimate, I also found some comments in substrate that suggest we shouldn't use `String` in runtime:
https://github.com/paritytech/polkadot-sdk/blob/9e403629d590abcd57c13f4010eb0714a4ee6642/substrate/frame/contracts/primitives/src/lib.rs#L67-L68
https://github.com/paritytech/polkadot-sdk/blob/9e403629d590abcd57c13f4010eb0714a4ee6642/substrate/primitives/std/without_std.rs#L43-L45

Reply:
I would argue we should use `sp-std` less, ideally not use at all and just write idiomatic Rust code, which in this case means just using `alloc::string::String`. So many changes during Substrate upgrades are due to re-exports as well.